File: wpimath/src/test/java/edu/wpi/first/math/kinematics/SwerveDriveKinematicsSerdeTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.kinematics;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.proto.Geometry2D;
import edu.wpi.first.math.proto.Kinematics.ProtobufSwerveDriveKinematics;
import org.junit.jupiter.api.Test;

public class SwerveDriveKinematicsSerdeTest {
  private static final SwerveDriveKinematics DATA =
      new SwerveDriveKinematics(
          new Translation2d(1, 2),
          new Translation2d(3, 4),
          new Translation2d(5, 6),
          new Translation2d(7, 8));

  @Test
  void testProtoPack() {
    ProtobufSwerveDriveKinematics proto = SwerveDriveKinematics.proto.createMessage();
    SwerveDriveKinematics.proto.pack(proto, DATA);

    assertEquals(DATA.getModules().length, proto.getModules().length());
    for (int i = 0; i < DATA.getModules().length; ++i) {
      assertEquals(DATA.getModules()[i].getX(), proto.getModules().get(i).getXMeters());
      assertEquals(DATA.getModules()[i].getY(), proto.getModules().get(i).getYMeters());
    }
  }

  @Test
  void testProtoUnpack() {
    ProtobufSwerveDriveKinematics proto = SwerveDriveKinematics.proto.createMessage();
    for (Translation2d translation : DATA.getModules()) {
      Geometry2D.ProtobufTranslation2d protoTranslation =
          Geometry2D.ProtobufTranslation2d.newInstance();
      Translation2d.proto.pack(protoTranslation, translation);
      proto.addModules(protoTranslation);
    }

    SwerveDriveKinematics data = SwerveDriveKinematics.proto.unpack(proto);
    assertArrayEquals(DATA.getModules(), data.getModules());
  }
}

File: wpimath/src/test/native/cpp/kinematics/serde/SwerveDriveKinematicsSerdeTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "frc/geometry/SwerveDriveKinematics.h"
#include "geometry2d.pb.h"

using namespace frc;

namespace {

using StructType = wpi::Struct<frc::SwerveDriveKinematics>;
using ProtoType = wpi::Protobuf<frc::SwerveDriveKinematics>;

constexpr std::array<uint8_t, StructType::kSize> create_test_buffer() {
  std::array<uint8_t, StructType::kSize> output;
  int buffer[] = {
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
};
  for (size_t idx = 0; idx < StructType::kSize; ++idx) {
    output[idx] = static_cast<uint8_t>(buffer[idx]);
  }
  return output;
}

std::array<uint8_t, StructType::kSize> kExpectedStructBuffer =
    create_test_buffer();

constexpr SwerveDriveKinematics kExpectedData{};
}  // namespace

TEST(SwerveDriveKinematicsTest, StructPack) {
  uint8_t buffer[StructType::kSize];
  StructType::Pack(buffer, kExpectedData);

  for (size_t i = 0; i < StructType::kSize; ++i) {
    EXPECT_EQ(kExpectedStructBuffer[i], buffer[i]) << " on byte " << i;
  }
}

TEST(SwerveDriveKinematicsTest, StructUnpack) {
  SwerveDriveKinematics unpacked_data = StructType::Unpack(kExpectedStructBuffer);

    EXPECT_EQ(kExpectedData.Modules(), unpacked_data.Modules());
}


TEST(SwerveDriveKinematicsTest, ProtobufPack) {
  wpi::proto::ProtobufSwerveDriveKinematics proto;
  ProtoType::Pack(&proto, kExpectedData);

    EXPECT_EQ(kExpectedData.Modules(), proto.modules());
}



TEST(SwerveDriveKinematicsTest, ProtobufUnpack) {
  wpi::proto::ProtobufSwerveDriveKinematics proto;
  proto.mutable_modules()->(kExpectedData.Modules());

  SwerveDriveKinematics unpacked_data = ProtoType::Unpack(proto);
    EXPECT_EQ(kExpectedData.Modules(), unpacked_data.Modules());
}